File: uikit/src/main/java/screen/call/MissedCall.java
package screen.call;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.cometchat.pro.constants.CometChatConstants;
import com.cometchat.pro.core.Call;
import com.cometchat.pro.core.CometChat;
import com.cometchat.pro.core.MessagesRequest;
import com.cometchat.pro.exceptions.CometChatException;
import com.cometchat.pro.models.BaseMessage;
import com.cometchat.pro.models.Group;
import com.cometchat.pro.models.User;
import com.cometchat.pro.uikit.CometChatCallList;
import com.cometchat.pro.uikit.R;
import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import constant.StringContract;
import listeners.OnItemClickListener;
import screen.CometChatCallActivity;
import screen.CometChatGroupDetailScreenActivity;
import screen.CometChatUserDetailScreenActivity;
import utils.Utils;

/**
 * MissedCall.class is a Fragment class used to display all the call that has been missed by loggedIn
 * user.
 *
 * Created At : 25th March 2020
 *
 * Modified On : 02nd April 2020
 */
public class MissedCall extends Fragment {

    private CometChatCallList rvCallList;

    private LinearLayout noCallView;

    private MessagesRequest messagesRequest;

    private LinearLayoutManager linearLayoutManager;

    private String loggedInUid = CometChat.getLoggedInUser().getUid();

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_missed_call, container, false);
        rvCallList = view.findViewById(R.id.callList_rv);
        linearLayoutManager = new LinearLayoutManager(getContext(),LinearLayoutManager.VERTICAL,false);
        rvCallList.setLayoutManager(linearLayoutManager);
        noCallView = view.findViewById(R.id.no_call_vw);
        rvCallList.setItemClickListener(new OnItemClickListener<Call>() {
            @Override
            public void OnItemClick(Call var, int position) {
                if (var.getReceiverType().equals(CometChatConstants.RECEIVER_TYPE_USER)) {
                    User user;
                    if (((User)var.getCallInitiator()).getUid().equals(CometChat.getLoggedInUser().getUid())) {
                        user =  ((User)var.getCallReceiver());
                    }
                    else
                    {
                        user = (User)var.getCallInitiator();
                    }
                    Intent intent = new Intent(getContext(), CometChatUserDetailScreenActivity.class);
                    intent.putExtra(StringContract.IntentStrings.UID, user.getUid());
                    intent.putExtra(StringContract.IntentStrings.NAME, user.getName());
                    intent.putExtra(StringContract.IntentStrings.AVATAR, user.getAvatar());
                    intent.putExtra(StringContract.IntentStrings.STATUS, user.getStatus());
                    intent.putExtra(StringContract.IntentStrings.IS_BLOCKED_BY_ME, user.isBlockedByMe());
                    intent.putExtra(StringContract.IntentStrings.FROM_CALL_LIST,true);
                    startActivity(intent);
                }
                else {
                    Group group;
                    group = ((Group)var.getCallReceiver());
                    Intent intent = new Intent(getContext(), CometChatGroupDetailScreenActivity.class);
                    intent.putExtra(StringContract.IntentStrings.GUID, group.getGuid());
                    intent.putExtra(StringContract.IntentStrings.NAME, group.getName());
                    intent.putExtra(StringContract.IntentStrings.AVATAR, group.getIcon());
                    intent.putExtra(StringContract.IntentStrings.MEMBER_SCOPE, group.getScope());
                    intent.putExtra(StringContract.IntentStrings.GROUP_OWNER, group.getOwner());
                    intent.putExtra(StringContract.IntentStrings.GROUP_TYPE,group.getGroupType());
                    intent.putExtra(StringContract.IntentStrings.MEMBER_COUNT,group.getMembersCount());
                    intent.putExtra(StringContract.IntentStrings.GROUP_PASSWORD,group.getPassword());
                    intent.putExtra(StringContract.IntentStrings.GROUP_DESC,group.getDescription());
                    startActivity(intent);
                }
            }
        });

        rvCallList.setItemCallClickListener(new OnItemClickListener<Call>() {
            @Override
            public void OnItemClick(Call var, int position) {
                checkOnGoingCall(var);
            }
        });
        rvCallList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                if (!rvCallList.canScrollVertically(1))
                {
                    getCallList();
                }
            }
        });
        return view;
    }

    private void checkOnGoingCall(Call var) {
        if(CometChat.getActiveCall()!=null && CometChat.getActiveCall().getCallStatus().equals(CometChatConstants.CALL_STATUS_ONGOING) && CometChat.getActiveCall().getSessionId()!=null) {
            AlertDialog.Builder alert = new AlertDialog.Builder(getContext());
            alert.setTitle(getContext().getResources().getString(R.string.ongoing_call))
                    .setMessage(getContext().getResources().getString(R.string.ongoing_call_message))
                    .setPositiveButton(getContext().getResources().getString(R.string.join), new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            Utils.joinOnGoingCall(getContext());
                        }
                    }).setNegativeButton(getContext().getResources().getString(R.string.cancel), new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    dialog.dismiss();
                }
            }).create().show();
        }
        else {
            initiateCall(var);
        }
    }

    private void initiateCall(Call var) {
        CometChat.initiateCall(var, new CometChat.CallbackListener<Call>() {
            @Override
            public void onSuccess(Call call) {
                Log.e("onSuccess: ", call.toString());
                if (call.getReceiverType().equals(CometChatConstants.RECEIVER_TYPE_USER)) {
                    User user;
                    if (((User) call.getCallInitiator()).getUid().equals(CometChat.getLoggedInUser().getUid())) {
                        user = ((User) call.getCallReceiver());
                    } else {
                        user = (User) call.getCallInitiator();
                    }
                    Utils.startCallIntent(getContext(), user, CometChatConstants.CALL_TYPE_AUDIO, true, call.getSessionId());
                } else
                    Utils.startGroupCallIntent(getContext(), ((Group) call.getCallReceiver()), CometChatConstants.CALL_TYPE_AUDIO, true, call.getSessionId());
            }

            @Override
            public void onError(CometChatException e) {
                if (rvCallList != null)
                    Snackbar.make(rvCallList, e.getMessage(), Snackbar.LENGTH_LONG).show();
            }
        });
    }

    private void getCallList() {
        if (messagesRequest == null)
        {
            messagesRequest = new MessagesRequest.MessagesRequestBuilder().setCategory(CometChatConstants.CATEGORY_CALL).build();
        }

        messagesRequest.fetchPrevious(new CometChat.CallbackListener<List<BaseMessage>>() {
            @Override
            public void onSuccess(List<BaseMessage> baseMessages) {
                Collections.reverse(baseMessages);
                rvCallList.setCallList(filteredMissedCall(baseMessages));
                if (rvCallList.size()!=0)
                {
                    noCallView.setVisibility(View.GONE);
                }
                else
                    noCallView.setVisibility(View.VISIBLE);
            }

            @Override
            public void onError(CometChatException e) {
                if (rvCallList!=null)
                    Snackbar.make(rvCallList,R.string.call_list_error,Snackbar.LENGTH_LONG).show();
            }
        });
    }

    private List<BaseMessage> filteredMissedCall(List<BaseMessage> baseMessages) {
        List<BaseMessage> filteredList = new ArrayList<>();
        for (BaseMessage baseMessage : baseMessages) {
            Call call = (Call)baseMessage;
            if(!((User)call.getCallInitiator()).getUid().equals(loggedInUid)) {
                if (((Call) baseMessage).getCallStatus().equals(CometChatConstants.CALL_STATUS_UNANSWERED)
                        || ((Call) baseMessage).getCallStatus().equals(CometChatConstants.CALL_STATUS_CANCELLED)) {
                    filteredList.add(baseMessage);
                }
            }
        }
        return filteredList;
    }

    @Override
    public void setMenuVisibility(boolean menuVisible) {
        super.setMenuVisibility(menuVisible);
        if (menuVisible)
        {
            getCallList();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
    }
}

File: uikit-kotlin/src/main/java/com/cometchat/pro/uikit/ui_components/calls/call_list/MissedCall.kt
package com.cometchat.pro.uikit.ui_components.calls.call_list

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.cometchat.pro.constants.CometChatConstants
import com.cometchat.pro.core.Call
import com.cometchat.pro.core.CometChat
import com.cometchat.pro.core.CometChat.CallbackListener
import com.cometchat.pro.core.MessagesRequest
import com.cometchat.pro.core.MessagesRequest.MessagesRequestBuilder
import com.cometchat.pro.exceptions.CometChatException
import com.cometchat.pro.models.BaseMessage
import com.cometchat.pro.models.Group
import com.cometchat.pro.models.User
import com.cometchat.pro.uikit.ui_components.shared.cometchatCalls.CometChatCalls
import com.cometchat.pro.uikit.R
import com.google.android.material.snackbar.Snackbar
import com.cometchat.pro.uikit.ui_resources.constants.UIKitConstants
import com.cometchat.pro.uikit.ui_resources.utils.item_clickListener.OnItemClickListener
import com.cometchat.pro.uikit.ui_components.groups.group_detail.CometChatGroupDetailActivity
import com.cometchat.pro.uikit.ui_components.users.user_details.CometChatUserDetailScreenActivity
import com.cometchat.pro.uikit.ui_resources.utils.ErrorMessagesUtils
import com.cometchat.pro.uikit.ui_resources.utils.Utils
import java.util.*

/**
 * MissedCall.class is a Fragment class used to display all the call that has been missed by loggedIn
 * user.
 *
 * Created At : 25th March 2020
 *
 * Modified On : 02nd April 2020
 */
class MissedCall : Fragment() {
    private var rvCallList: CometChatCalls? = null
    private var noCallView: LinearLayout? = null
    private var messagesRequest: MessagesRequest? = null
    private var linearLayoutManager: LinearLayoutManager? = null
    private val loggedInUid = CometChat.getLoggedInUser().uid
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_cometchat_missed_call, container, false)
        rvCallList = view.findViewById(R.id.callList_rv)
        linearLayoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        rvCallList!!.setLayoutManager(linearLayoutManager)
        noCallView = view.findViewById(R.id.no_call_vw)
        rvCallList!!.setItemClickListener(object : OnItemClickListener<Call?>() {
            override fun OnItemClick(t: Any, position: Int) {
                var call = t as Call
                if (call.receiverType == CometChatConstants.RECEIVER_TYPE_USER) {
                    val user: User
                    user = if ((call.callInitiator as User).uid == CometChat.getLoggedInUser().uid) {
                        call.callReceiver as User
                    } else {
                        call.callInitiator as User
                    }
                    val intent = Intent(context, CometChatUserDetailScreenActivity::class.java)
                    intent.putExtra(UIKitConstants.IntentStrings.UID, user.uid)
                    intent.putExtra(UIKitConstants.IntentStrings.NAME, user.name)
                    intent.putExtra(UIKitConstants.IntentStrings.AVATAR, user.avatar)
                    intent.putExtra(UIKitConstants.IntentStrings.STATUS, user.status)
                    intent.putExtra(UIKitConstants.IntentStrings.IS_BLOCKED_BY_ME, user.isBlockedByMe)
                    intent.putExtra(UIKitConstants.IntentStrings.FROM_CALL_LIST, true)
                    startActivity(intent)
                } else {
                    val group: Group
                    group = call.callReceiver as Group
                    val intent = Intent(context, CometChatGroupDetailActivity::class.java)
                    intent.putExtra(UIKitConstants.IntentStrings.GUID, group.guid)
                    intent.putExtra(UIKitConstants.IntentStrings.NAME, group.name)
                    intent.putExtra(UIKitConstants.IntentStrings.AVATAR, group.icon)
                    intent.putExtra(UIKitConstants.IntentStrings.MEMBER_SCOPE, group.scope)
                    intent.putExtra(UIKitConstants.IntentStrings.GROUP_OWNER, group.owner)
                    intent.putExtra(UIKitConstants.IntentStrings.GROUP_TYPE, group.groupType)
                    intent.putExtra(UIKitConstants.IntentStrings.MEMBER_COUNT, group.membersCount)
                    intent.putExtra(UIKitConstants.IntentStrings.GROUP_PASSWORD, group.password)
                    intent.putExtra(UIKitConstants.IntentStrings.GROUP_DESC, group.description)
                    startActivity(intent)
                }
            }
        })
        rvCallList!!.setItemCallClickListener(object : OnItemClickListener<Call?>() {
            override fun OnItemClick(t: Any, position: Int) {
                checkOnGoingCall(t as Call)
            }
        })
        rvCallList!!.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                if (!rvCallList!!.canScrollVertically(1)) {
                    callList
                }
            }
        })
        return view
    }

    private fun checkOnGoingCall(`var`: Call) {
        if (CometChat.getActiveCall() != null && CometChat.getActiveCall().callStatus == CometChatConstants.CALL_STATUS_ONGOING && CometChat.getActiveCall().sessionId != null) {
            val alert = AlertDialog.Builder(context)
            alert.setTitle(context!!.resources.getString(R.string.ongoing_call))
                    .setMessage(context!!.resources.getString(R.string.ongoing_call_message))
                    .setPositiveButton(context!!.resources.getString(R.string.join)) { dialog, which -> Utils.joinOnGoingCall(context!!) }.setNegativeButton(context!!.resources.getString(R.string.cancel)) { dialog, which -> dialog.dismiss() }.create().show()
        } else {
            initiateCall(`var`)
        }
    }

    private fun initiateCall(`var`: Call) {
        CometChat.initiateCall(`var`, object : CallbackListener<Call>() {
            override fun onSuccess(call: Call) {
                Log.e("onSuccess: ", call.toString())
                if (call.receiverType == CometChatConstants.RECEIVER_TYPE_USER) {
                    val user: User
                    user = if ((call.callInitiator as User).uid == CometChat.getLoggedInUser().uid) {
                        call.callReceiver as User
                    } else {
                        call.callInitiator as User
                    }
                    Utils.startCallIntent(context!!, user, CometChatConstants.CALL_TYPE_AUDIO, true, call.sessionId)
                } else Utils.startGroupCallIntent(context!!, call.callReceiver as Group, CometChatConstants.CALL_TYPE_AUDIO, true, call.sessionId)
            }

            override fun onError(e: CometChatException) {
                if (rvCallList != null) ErrorMessagesUtils.cometChatErrorMessage(context, e.code)
            }
        })
    }

    private val callList: Unit
        private get() {
            if (messagesRequest == null) {
                messagesRequest = MessagesRequestBuilder().setCategories(listOf(CometChatConstants.CATEGORY_CALL)).build()
            }
            messagesRequest!!.fetchPrevious(object : CallbackListener<List<BaseMessage?>>() {
                override fun onSuccess(baseMessages: List<BaseMessage?>) {
                    Collections.reverse(baseMessages)
                    rvCallList!!.setCallList(filteredMissedCall(baseMessages))
                    if (rvCallList!!.size() != 0) {
                        noCallView!!.visibility = View.GONE
                    } else noCallView!!.visibility = View.VISIBLE
                }

                override fun onError(e: CometChatException) {
                    if (rvCallList != null)
                        ErrorMessagesUtils.showCometChatErrorDialog(context, resources.getString(R.string.call_list_error),UIKitConstants.ErrorTypes.ERROR)
                }
            })
        }

    private fun filteredMissedCall(baseMessages: List<BaseMessage?>): List<BaseMessage?> {
        val filteredList: MutableList<BaseMessage?> = ArrayList()
        for (baseMessage in baseMessages) {
            val call = baseMessage as Call?
            if ((call!!.callInitiator as User).uid != loggedInUid) {
                if (call.callStatus == CometChatConstants.CALL_STATUS_UNANSWERED || call.callStatus == CometChatConstants.CALL_STATUS_CANCELLED) {
                    filteredList.add(baseMessage)
                }
            }
        }
        return filteredList
    }

    override fun setMenuVisibility(menuVisible: Boolean) {
        super.setMenuVisibility(menuVisible)
        if (menuVisible) {
            callList
        }
    }

    override fun onResume() {
        super.onResume()
    }
}